### DISCUSSION TITLE
Executar código Dart em tempo de execução

### DISCUSSION THREAD

**Discussion by rubensgalvan** [Flutterando/forum#38] [Discussões gerais]
Existe alguma maneira de executar um código dart, em tempo de execução, igual o DartPad faz?

Tipo:

```
runCode.source = "return Text('Exemplo')";
Widget text = <Widget> runCode.run();
```

Um exemplo simples, não demonstra a real utilidade que atualmente faço uso, mas poderia ser mais complexo.

Tenho um sistema em Java que tem um interpretador Java in Runtime. Algumas coisas em clientes, não precisa compilar uma nova versão, poderia ser uma configuração no próprio cliente. Altera o código no cliente, fechou e abriu a tela, já esta funcionando.

**Comment by eduardoflorence** [comment]
O DartPad não executa o código em tempo de execução, é necessário que o código Dart seja compilado.
O que o Dartpad faz é entregar o código para um servidor e recebê-lo de volta compilado.
Tanto o Dartpad, quanto este servidor são de código aberto e abaixo estão os repositórios para você conferir melhor:
**Dartpad**: https://github.com/dart-lang/dart-pad
**Servidor**: https://github.com/dart-lang/dart-services

Acredito que uma alternativa para o que vocês está querendo é usar um recurso que estão chamando de `Flutter Server Driven UI`. Você pode entender melhor lendo este artigo: https://danielllewellyn.medium.com/in-this-article-im-going-to-demonstrate-a-way-of-driving-the-ui-of-flutter-from-the-backend-43478b46e202
Existe também um package que avançou bem neste tema: https://pub.dev/packages/sdui

**Comment by mateusfccp** [comment]
O Dart não possui um método como `eval`.

Dependendo do seu caso, no entanto, você pode tentar usar o [`Isolate.spawnUri`](https://api.flutter.dev/flutter/dart-isolate/Isolate/spawnUri.html). Você consegue passar uma `Uri` que represente un método `main`, e este será executado em uma nova _isolate_.

Você vai ter que entender um pouco de como funcionam _isolates_ e como se comunicar com ela.

**Comment by jacobaraujo7** [comment]
Também tem algo que estou estudando e até consigo fazer funcionar.
https://pub.dev/packages/dart_eval

**Reply by rubensgalvan** [reply]
Bem isso ai, o flutter_eval aparenta ser mais completo, podendo gerar Widgets e criar algumas regras bem específicas.

Este tipo de arquitetura, não é muito ortodoxa no mundo dos Devs. Mas eu vejo no dia a dia o quanto este tipo de recurso diminui drasticamente o tempo de resposta a bugs e correções específicas em clientes em Runtime, sem precisar criar uma sprint, gerar atividades, alterar o bug, gerar uma release, manual, disponibilizar para o cliente, e o cliente atualizar, custa um tempo gigante, onde 10 minutos, resolvem para o cliente, e ai pode ser criada na software house o processo correto pra correção e atualização.

**Reply by rubensgalvan** [reply]
Perfeito, o Flutter e Dart, terem um embrião deste tipo de recurso, para quem quiser usar esta maravilha.
